permuta/misc/iterable_floor_and_ceiling.py

```diff
@@ -0,0 +1,62 @@
+import collections
+
+FloorAndCeiling = collections.namedtuple("FloorAndCeiling", ["floor", "ceiling"])
+
+def left_floor_and_ceiling(iterable, default_floor=None, default_ceiling=None):
+    """Find the left floor and ceiling indices of iterable.
+
+    Define left_floor of an element in a sequence to be the index of the
+    greatest smaller element to the left of said element, or default_floor
+    if there is none. Similarly define default_ceiling. This function yields
+    a tuple (left_floor, left_ceiling) for each element of iterable.
+
+    Args:
+        iterable: <iterable>
+            An iterable of totally ordered unique elements.
+        default_floor: <object>
+        default_ceiling: <object>
+
+    Yields: (int, int)
+        The i-th yielded tuple is the left floor and ceiling indices of
+        the i-th element of the iterable. The tuples are named tuples
+        with floor and ceiling attributes.
+    """
+    # TODO: Define behaviour for duplicate elements
+    dq = collections.deque()
+    smallest = None
+    biggest = None
+    index = 0
+    for element in iterable:
+        if index is 0:
```

Comment:
We were actually discussing this earlier!
Thanks for noticing, I'll get right on it.